src/main/scala/is/hail/expr/ir/Compile.scala

```diff
@@ -98,3 +110,121 @@ object Compile {
     ), body)
   }
 }
+
+object CompileWithAggregators {
+  def apply[
+    F1 >: Null : TypeInfo,
+    F2 >: Null : TypeInfo,
+    R: TypeInfo : ClassTag
+  ](aggName: String,
+    aggType: TAggregable,
+    args: Seq[(String, Type, ClassTag[_])],
+    aggScopeArgs: Seq[(String, Type, ClassTag[_])],
+    body: IR
+  ): (Array[RegionValueAggregator], Array[() => F1], Type, () => F2, Type) = {
+
+    assert((args ++ aggScopeArgs).forall { case (_, t, ct) => TypeToIRIntermediateClassTag(t) == ct })
+
+    // the aggregable must be argument 0 to post-aggregated IR, so must have a place holder in env
+    val env = ((aggName, aggType, TypeToIRIntermediateClassTag(aggType)) +: args)
+      .zipWithIndex
+      .foldLeft(new Env[IR]()) { case (e, ((n, t, _), i)) => e.bind(n, In(i, t)) }
+
+    val ir = Subst(body, env)
+    Infer(ir, Some(aggType))
+
+    val (postAggIR, aggResultType, aggregators) = ExtractAggregators(ir, aggType)
+
+    val f1TypeInfo: Array[GenericTypeInfo[_]] =
+      GenericTypeInfo[Region]() +:
+        GenericTypeInfo[RegionValueAggregator]() +:
+        (aggType +: aggScopeArgs.map(_._2).toArray).flatMap { t =>
+          List[GenericTypeInfo[_]](GenericTypeInfo()(typeToTypeInfo(t)), GenericTypeInfo[Boolean]()).iterator
+        }
+
+    val (rvAggs, seqOps) = aggregators.zipWithIndex.map { case ((ir, agg), i) =>
+      val fb = new FunctionBuilder[F1](f1TypeInfo.asInstanceOf[Array[MaybeGenericTypeInfo[_]]], GenericTypeInfo[Unit]())
+      Emit(ir, fb, 2, aggType)
+      (agg, fb.result())
+    }.unzip
+
+    val args1 = ("AGGR", aggResultType, classTag[Long]) +: args
+    val (t, f) = Compile[F2, R](args1, postAggIR)
```

Comment:
Hmm. This isn't quite how I had first imagined this setup. `Compile` sort of packages up a bunch of stuff that I was doing manually with `Emit`. It kind of puns on something: these `args1` indicate both:
 - how many parameters the resulting scala lambda, `f`, will have, and
 - some names (e.g. `globals`, `AGGR`) that should be replaced with references to inputs (e.g. `In(0)`).

This is what you want in most cases where the user references, say, `sa`. However, in this case, `ExtractAggregators` actually replaces `ApplyAggOp`s with a `In(0, ...)` (not with a `Ref("AGGR")`). The `postAggIR` has no references to `AGGR` because they always appear deeper than an `ApplyAggOp`. However, this puts you in the awkward situation having to give `Compile` enough arguments so that it creates the right arity function.

I don't like this set up because we do a `Subst` to replace `AGGR`, but `AGGR` should never be there anyway. Seems dangerous to do something we expect to be a no-op (but don't check is a no-op). Maybe a solution is to change `ExtractAggregators` to replace `ApplyAggOp` with a `Ref(aggname)`. This would change several tests though.

What do you think?